src/Zone.js

```diff
@@ -18,20 +20,77 @@ export default class Zone extends UXBase {
 	// Name used for UserTiming measures
 	measureName = this.props.name;
 
-	// Create a new `ExpectedMark` for each mark
-	marks = this.props.marks.map(markName => {
-		const mark = ExpectedMark.create(markName);
+	constructor(props) {
+		super(props);
+
+		// Create a new `ExpectedMark` for each mark
+		const configuredMarkNames = this.props.elements
+			? // new elements array on zone object
+			  this.props.elements.map(element => element.marks).flat()
+			: // legacy with direct marks array on zone object
+			  this.props.marks;
 
-		mark.onComplete(completeMark => {
-			// pass the event upstream
-			this.props.onMark(markName);
-			if (this.marks.every(m => m.marked)) {
-				this.measure(markName);
+		// mark names for elements already on the page
+		let alreadyOnThePage = [];
+
+		if (this.props.startMarkName === INTERACTIVE_TRANSITION_START_MARK_NAME) {
+			// if mark selector is configured, use it to find marks to record
+			if (props.markSelector) {
+				alreadyOnThePage.concat(
+					configuredMarkNames.filter(
+						markName => props.markSelector(markName).length > 0
+					)
+				);
 			}
+
+			// if elements have selectors defined or global element selector is configured,
+			// use them to find marks to record
+			props.elements.forEach(element => {
```

Comment:
instead of relying on side effects to build the `alreadyOnThePage`, consider using `.reduce` to make this a little easier to read and maintain, e.g.

```js
const alreadyOnThePage = props.elements.reduce((acc, element) => {
  const elementSelector = element.elementSelector || this.props.elementSelector;
  const selected = typeof elementSelector === 'function'
    ? elementSelector()
    : document.querySelectorAll(elementSelector);
  if (selected.length > 0) {
    return acc.concat(element.marks);
  }
  return acc;
}, []);
```